ckan/model/user.py

```diff
@@ -332,7 +333,21 @@ def user_ids_for_name_or_id(cls, user_list: Iterable[str]=()) -> list[str]:
                                  cls.id.in_(user_list)))  # type: ignore
         return [user.id for user in query.all()]
 
-    def set_user_last_active(self):
+    def get_id(self) -> str:
+        '''Needed by flask-login'''
+        return self.id
+
+    @property
+    def is_authenticated(self) -> bool:
+        '''Needed by flask-login'''
+        return True
+
+    @property
+    def is_anonymous(self):
+        '''Needed by flask-login'''
+        return False
+
```

Comment:
Our `User` class inherits from `StatefulObjectMixin`, which has a `is_active()` function. But Flask-login requires it to be a [property](https://flask-login.readthedocs.io/en/latest/#your-user-class), so we need to override it or otherwise it will always return `True` when it is evaluated:

```diff
diff --git a/ckan/model/user.py b/ckan/model/user.py
index 61d65f997..9d3c5728a 100644
--- a/ckan/model/user.py
+++ b/ckan/model/user.py
@@ -347,6 +347,12 @@ class User(core.StatefulObjectMixin,
         '''Needed by flask-login'''
         return False
 
+    @property
+    def is_active(self):
+        '''Needed by flask-login'''
+        return super().is_active()
+
+
     def set_user_last_active(self) -> None:
         if self.last_active:
             session["last_active"] = self.last_active
```

We should probably document this change in the CHANGELOG

Reply:
flask-login uses it only in one place, [here](https://github.com/maxcountryman/flask-login/blob/bb254077ac98ad4bd0ca15f5b6818a3d8f8b5dd7/src/flask_login/utils.py#L193-L194), and it useses it for one purpose only, if the user is not active it won't let the user log in unless you pass force=True. 
But we have our custom check [here](https://github.com/ckan/ckan/pull/6560/files#diff-bced5766e0c0f50671e5e0f57217ef7e79f2496ed3b2c711f90daeef25dccdf1) in our custom authenticator that checks for user inactivity and will return None instead of user_obj, that means the login_user() in this situation is never reached as it will fail to the else block
![image](https://user-images.githubusercontent.com/72216462/175969734-98a20095-f806-4a8e-8681-ac6c1beaa438.png)


@amercader please let me know how do you want me to proceed 

Thanks

Reply:
I can make these changes, remove that check in our custom authenticator and let the flask-login to handle this, then, we could end up with something like this:
```python
if login_user(user_obj):
    return next_page_or_default(next)
else:
     err = _('Login as {} failed - user isn\'t active'.format(username_or_email))
     h.flash_error(err)
     return base.render("user/login.html", extra_vars)
```